### ISSUE TITLE
3D Image Classification

### ISSUE THREAD

**Issue by JBarsotti** [shap/shap#3097]
### Problem Description

I am trying to use DeepExplainer for a convolutional neural network that was trained to identify brain images. These are 3D brain images that are divided into slices. The shape of the image is (1, 140, 140, 140), where the third dimension is the number of slices (140).

Is there any way to use SHAP in the case of a 3D image such as this? I have attempted to implement it using the method suggested in the [documentation](https://shap-lrjball.readthedocs.io/en/latest/example_notebooks/deep_explainer/PyTorch%20Deep%20Explainer%20MNIST%20example.html):

batch = next(iter(test_loader))
images, _ = batch

background = images[:100]
test_images = images[100:103]

e = shap.DeepExplainer(model, background)
shap_values = e.shap_values(test_images)

shap_numpy = [np.swapaxes(np.swapaxes(s, 1, -1), 1, 2) for s in shap_values]
test_numpy = np.swapaxes(np.swapaxes(test_images.numpy(), 1, -1), 1, 2)

shap.image_plot(shap_numpy, -test_numpy)

However, this example is for MNIST 2D images, not 3D.

### Feature request checklist

- [X] I have checked the issue tracker for duplicate issues.
- [ ] I'd be interested in making a PR to implement this feature

**Comment by connortann** [comment]
I think this is perfectly possible, although it may be significantly more computationally expensive. The `DeepExplainer` accepts tensors of any shape, providing the tensor matches what is expected by the model.

What have you tried so far: have you encountered a specific issue, or are you currently just tentatively exploring ideas?

**Comment by JBarsotti** [comment]
Thank you for the reply! So far, I have developed some code that runs, but I have no idea whether it is doing the right thing. I plotted the results as a heatmap because I don't know how to use shap.image_plot(). I also took the slice with maximum SHAP value (again, don't know if this is justifiable). A further issue that has arisen is the fact that I don't think SHAP recognizes the neural network I'm using because I see these warnings:

Warning: unrecognized nn.Module: Flatten
Warning: unrecognized nn.Module: _DenseBlock
Warning: unrecognized nn.Module: _Transition

I don't know if that messes up the SHAPLEY algorithm or not. Anyway, here is the code below. I borrowed some of it from a research group nice enough to share their saliency mapping code.

  device = torch.device("cuda")

  state_dict = convert_state_dict("ABCDt2_15-07-23-18_32.pt")
  net        = DenseNet(3,1,1)
  net.load_state_dict(state_dict)
  net        = net.to(device)

for k, data in enumerate(tqdm.tqdm(back_loader)):
    
    im, age = data
    im      = next(iter(im))
    
    im = torch.tensor(im).to(device = device).view(1, 1, 140, 140, 140)
    e  = shap.DeepExplainer(net, im)
    
for k, data in enumerate(tqdm.tqdm(test_loader)):    

    im, age = data
    im      = next(iter(im))
    
    im = torch.tensor(im).to(device = device).view(1, 1, 140, 140, 140)
    shap_values = e.shap_values(im)
    
    squeezed    = shap_values.squeeze()
    max_arg_max = np.argmax(np.max(squeezed, axis = (0, 1))) 
    max_slice = squeezed[:, :, max_arg_max]

    max_slice[(max_slice < 0.0001) & (max_slice > -0.0001)] = np.nan
    fig = plt.figure()
    ax  = fig.subplots()
    ax  = sns.heatmap(max_slice, annot = False, fmt = 'd', linewidths = 0, cmap = 'viridis')
    ax.tick_params(left = False, bottom = False)
    ax.set_ylabel(None)
    ax.set_xlabel(None)
    ax.set_yticks([])
    ax.set_xticks([])
    plt.show()

**Comment by connortann** [comment]
Could you provide a [minimal reproducible example](https://en.wikipedia.org/wiki/Minimal_reproducible_example)?

**Comment by JBarsotti** [comment]
Maybe ... I can't provide the data I used because it is restricted access. I'll need to find an open source brain image.

**Comment by connortann** [comment]
For a minimal example, it would be better to use a minimal dummy dataset, that is reproducible from the code alone. 

You could try something like:

```python
np.random_state(0)
data = np.random.random(size=(3,3,3))
```

**Comment by JBarsotti** [comment]
Sorry for not sending an example yet. I'm still trying to figure out the best way. The program predicts the age of a brain based on anatomical features in the image, so I'm not yet sure how best to give a dummy dataset. There are a few open source datasets I know of.

**Comment by connortann** [comment]
For a toy example, the actual content of the data isn't really relevant - it doesn't have to be at all realistic if we just need to establish whether `shap` will run on a particular kind of model. A very small synthetic example will do fine.

I think we have a few different discussion points:

### 1. Does DeepExplainer work on 3D tensors?

I think shap will work just fine with 3D tensors - you will get Shapley values with the same shape as your input data, and it's up to you how to visualise them.

### 2.  How can one summarise a 3D tensor of Shapley values?

Taking the maximum seems like a reasonable way to project from 3D to 2D. You could also try taking the average, which may be more robust to noise and outliers.

### 3. Handling unrecognised layers

I'm unsure about the warnings you mention. I think these layers might not be supported:

```
Warning: unrecognized nn.Module: Flatten
Warning: unrecognized nn.Module: _DenseBlock
Warning: unrecognized nn.Module: _Transition
```

Possibly related issue (currently unresolved):  #2434

**Comment by JBarsotti** [comment]
Do you think that I should not be using Shapley because it does not support the layers? I don't know how that affects the results, if at all.

**Comment by connortann** [comment]
I would certainly advise caution in interpreting the results for your specific model until #2434 is closed. In the meantime, you might be able to use the workaround in that thread to replace the flatten layer.

I'll close this issue for now as a duplicate; we'll use the other issue to track the feature request.
